# Курс по математическому анализу. 2й курс, астрономы. Дата начала разработки октябрь 2020

# Oбработка данных на python


# Чтение и группировка данных

In [1]:
import pandas as pd # импорт библиотеки Пандас

football = pd.read_csv('data_sf.csv')
display(football.head())

Unnamed: 0               Name  Age Nationality                 Club  \
0           0           L. Messi   31   Argentina         FC Barcelona   
1           1  Cristiano Ronaldo   33    Portugal             Juventus   
2           2          Neymar Jr   26      Brazil  Paris Saint-Germain   
3           3             De Gea   27       Spain    Manchester United   
4           4       K. De Bruyne   27     Belgium      Manchester City   

       Value    Wage Position  Crossing  Finishing  ...  Penalties  Composure  \
0  110500000  565000       RF        84         95  ...         75         96   
1   77000000  405000       ST        84         94  ...         85         95   
2  118500000  290000       LW        79         87  ...         81         94   
3   72000000  260000       GK        17         13  ...         40         68   
4  102000000  355000      RCM        93         82  ...         79         88   

   Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  GKKicking  \
0       33              28             26         6          11         15   
1       28              31             23         7          11         15   
2       27              24             33         9           9         15   
3       15              21             13        90          85         87   
4       68              58             51        15          13          5   

   GKPositioning  GKReflexes  
0             14           8  
1             14          11  
2             15          11  
3             88          94  
4             10          13  

[5 rows x 42 columns]

### Краткое описание файла метод .info()

In [3]:
display(football.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12897 entries, 0 to 12896
Data columns (total 42 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Unnamed: 0       12897 non-null  int64 
 1   Name             12897 non-null  object
 2   Age              12897 non-null  int64 
 3   Nationality      12897 non-null  object
 4   Club             12897 non-null  object
 5   Value            12897 non-null  int64 
 6   Wage             12897 non-null  int64 
 7   Position         12897 non-null  object
 8   Crossing         12897 non-null  int64 
 9   Finishing        12897 non-null  int64 
 10  HeadingAccuracy  12897 non-null  int64 
 11  ShortPassing     12897 non-null  int64 
 12  Volleys          12897 non-null  int64 
 13  Dribbling        12897 non-null  int64 
 14  Curve            12897 non-null  int64 
 15  FKAccuracy       12897 non-null  int64 
 16  LongPassing      12897 non-null  int64 
 17  BallControl      12897 non-null

None

### Подробное описание, метод .describe()

In [8]:
football.describe()

Unnamed: 0           Age         Value           Wage      Crossing  \
count  12897.000000  12897.000000  1.289700e+04   12897.000000  12897.000000   
mean    6605.040862     24.795379  2.112449e+06    7517.562224     47.076374   
std     3782.545526      4.872212  6.498424e+06   23061.985387     18.043470   
min        0.000000     16.000000  0.000000e+00    1000.000000      5.000000   
25%     3357.000000     21.000000  2.500000e+05    1000.000000     35.000000   
50%     6622.000000     24.000000  5.000000e+05    2000.000000     51.000000   
75%     9876.000000     28.000000  8.500000e+05    4000.000000     61.000000   
max    13125.000000     45.000000  1.185000e+08  565000.000000     93.000000   

          Finishing  HeadingAccuracy  ShortPassing       Volleys  \
count  12897.000000     12897.000000  12897.000000  12897.000000   
mean      42.876328        50.202528     56.026983     40.295030   
std       19.096935        17.281593     14.740436     17.080687   
min        2.000000         4.000000      7.000000      4.000000   
25%       28.000000        43.000000     50.000000     28.000000   
50%       45.000000        54.000000     60.000000     41.000000   
75%       59.000000        62.000000     65.000000     53.000000   
max       95.000000        94.000000     93.000000     90.000000   

          Dribbling  ...     Penalties     Composure       Marking  \
count  12897.000000  ...  12897.000000  12897.000000  12897.000000   
mean      52.461735  ...     46.459874     55.942932     45.655811   
std       18.871643  ...     15.479313     11.280631     19.456346   
min        4.000000  ...      5.000000     12.000000      3.000000   
25%       45.000000  ...     37.000000     49.000000     29.000000   
50%       58.000000  ...     47.000000     56.000000     51.000000   
75%       65.000000  ...     58.000000     63.000000     62.000000   
max       97.000000  ...     91.000000     96.000000     93.000000   

       StandingTackle  SlidingTackle      GKDiving    GKHandling  \
count    12897.000000   12897.000000  12897.000000  12897.000000   
mean        46.186710      44.386679     17.218345     16.962317   
std         21.172586      20.726546     18.085618     17.349624   
min          2.000000       3.000000      1.000000      1.000000   
25%         25.000000      23.000000      8.000000      8.000000   
50%         54.000000      51.000000     11.000000     11.000000   
75%         64.000000      62.000000     14.000000     14.000000   
max         93.000000      91.000000     90.000000     92.000000   

          GKKicking  GKPositioning    GKReflexes  
count  12897.000000   12897.000000  12897.000000  
mean      16.797938      16.950221     17.315965  
std       16.971411      17.369297     18.335817  
min        1.000000       1.000000      1.000000  
25%        8.000000       8.000000      8.000000  
50%       11.000000      11.000000     11.000000  
75%       14.000000      14.000000     14.000000  
max       91.000000      90.000000     94.000000  

[8 rows x 38 columns]

In [9]:
football.describe()['Age'] # Хотим выделить контректный столбец

count    12897.000000
mean        24.795379
std          4.872212
min         16.000000
25%         21.000000
50%         24.000000
75%         28.000000
max         45.000000
Name: Age, dtype: float64

football.describe()['Age']['mean'] # берем конкретное значение

#### Что делать если данные не числовые ? параметр  .describe(include = ['object'])



In [11]:
football.describe(include = ['object'])

Name Nationality             Club Position
count          12897       12897            12897    12897
unique         12326         156              650       27
top     J. Rodríguez     England  Shonan Bellmare       GK
freq               9        1368               30     1641

Здесь для каждой колонки с типом object (строчные данные) выводится количество непустых строк, уникальных значений, самое частое значение (top) и частота этого значения (freq). 

# Примеры

1. Каково среднее значение возраста футболистов в наборе данных (округлите до целого)?

In [19]:
import math 
aver_age = math.ceil(football.describe()['Age'] ['mean'])
print(aver_age)

25


2. Какое значение у первого (25%) квартиля параметра ShortPassing?

In [20]:
val_q1 = football.describe()['ShortPassing'] ['25%']
print(val_q1)

50.0


3. Какова самая частая национальность (Nationality) футболистов?

In [23]:
top_national = football.describe(include = ['object'])['Nationality']['top']
print(top_national)

England


4. Сколько разных клубов в наборе данных о футболистах?

In [26]:
unique_club = football.describe(include = ['object'])['Club']['unique']
print(unique_club)

650


# Извлекаем данные по условиям

Хотим выбрать только тех, у кого возраст больше 20 лет

In [27]:
football[football.Age > 20]

Unnamed: 0               Name  Age Nationality  \
0               0           L. Messi   31   Argentina   
1               1  Cristiano Ronaldo   33    Portugal   
2               2          Neymar Jr   26      Brazil   
3               3             De Gea   27       Spain   
4               4       K. De Bruyne   27     Belgium   
...           ...                ...  ...         ...   
12838       13067          D. Mackay   21    Scotland   
12857       13086             Wu Lei   21    China PR   
12861       13090        Y. Uchimura   33       Japan   
12870       13099           M. Hurst   22    Scotland   
12873       13102      K. Pilkington   44     England   

                             Club      Value    Wage Position  Crossing  \
0                    FC Barcelona  110500000  565000       RF        84   
1                        Juventus   77000000  405000       ST        84   
2             Paris Saint-Germain  118500000  290000       LW        79   
3               Manchester United   72000000  260000       GK        17   
4                 Manchester City  102000000  355000      RCM        93   
...                           ...        ...     ...      ...       ...   
12838                  Kilmarnock      30000    1000       GK        14   
12857         Tianjin Quanjian FC    9000000    2000      LDM        52   
12861  Hokkaido Consadole Sapporo      20000    1000      CAM        51   
12870            St. Johnstone FC      40000    1000       GK        19   
12873            Cambridge United          0    1000       GK        11   

       Finishing  ...  Penalties  Composure  Marking  StandingTackle  \
0             95  ...         75         96       33              28   
1             94  ...         85         95       28              31   
2             87  ...         81         94       27              24   
3             13  ...         40         68       15              21   
4             82  ...         79         88       68              58   
...          ...  ...        ...        ...      ...             ...   
12838         20  ...         24         30       12              18   
12857         29  ...         46         39       52              40   
12861         31  ...         33         62       45              48   
12870         15  ...         29         28       12              15   
12873         12  ...         22         56       15              15   

       SlidingTackle  GKDiving  GKHandling  GKKicking  GKPositioning  \
0                 26         6          11         15             14   
1                 23         7          11         15             14   
2                 33         9           9         15             15   
3                 13        90          85         87             88   
4                 51        15          13          5             10   
...              ...       ...         ...        ...            ...   
12838             18        48          50         47             48   
12857             49         8          15         11              8   
12861             42        13          15          7             15   
12870             16        45          49         50             50   
12873             13        45          48         44             49   

       GKReflexes  
0               8  
1              11  
2              11  
3              94  
4              13  
...           ...  
12838          48  
12857           5  
12861          15  
12870          45  
12873          46  

[10025 rows x 42 columns]

Возраст больше 20 и все из Японии

In [29]:
football[(football.Age > 20)&(football.Nationality=='Japan')]

Unnamed: 0          Name  Age Nationality                        Club  \
305           306       T. Inui   30       Japan                  Real Betis   
450           454   S. Nakajima   23       Japan             Portimonense SC   
490           494      H. Sakai   28       Japan      Olympique de Marseille   
527           531      K. Honda   32       Japan           Melbourne Victory   
635           643  G. Shibasaki   26       Japan                   Getafe CF   
...           ...           ...  ...         ...                         ...   
12577       12806   T. Nagaishi   22       Japan                Cerezo Osaka   
12623       12852   T. Nakagawa   25       Japan         Yokohama F. Marinos   
12737       12966     A. Suzuki   31       Japan         Yokohama F. Marinos   
12780       13009   M. Ichimaru   21       Japan                 Gamba Osaka   
12861       13090   Y. Uchimura   33       Japan  Hokkaido Consadole Sapporo   

          Value   Wage Position  Crossing  Finishing  ...  Penalties  \
305    14000000  30000       LM        77         70  ...         51   
450    14000000  12000       LW        66         73  ...         64   
490     8000000  44000       RB        74         33  ...         36   
527     8000000  12000       RM        75         65  ...         76   
635    11000000  24000       LM        68         70  ...         71   
...         ...    ...      ...       ...        ...  ...        ...   
12577     60000   1000       GK        12          7  ...         10   
12623     40000   1000       RW        40         61  ...         64   
12737     20000   1000       GK        11          9  ...         18   
12780     60000   1000      CDM        39         27  ...         37   
12861     20000   1000      CAM        51         31  ...         33   

       Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
305           76       65              39             31        11   
450           81       28              41             36         7   
490           68       77              77             80        13   
527           80       59              48             48         3   
635           75       61              73             72        10   
...          ...      ...             ...            ...       ...   
12577         33       17              12             12        56   
12623         52       34              10             11        10   
12737         31       13              10             14        49   
12780         47       47              39             51         7   
12861         62       45              48             42        13   

       GKHandling  GKKicking  GKPositioning  GKReflexes  
305            12         11             15           8  
450            10         10              8           6  
490             8         11             10           9  
527             3          5              5           3  
635            13         13              9           6  
...           ...        ...            ...         ...  
12577          46         44             51          53  
12623          11         11             13           9  
12737          48         54             52          54  
12780           9         10             12          14  
12861          15          7             15          15  

[347 rows x 42 columns]

Возраст больше 20 и все из Японии игроки на позициях GK. Каково у них значение Wage. Показать отсортированный ряд

In [38]:
nihonjin = football[(football.Age > 20)&(football.Nationality=='Japan')&(football.Position=='GK')]['Wage'].sort_values(ascending=False)
print(nihonjin)

1504     6000
2756     4000
1854     3000
2601     3000
5559     2000
9816     2000
9539     2000
8859     2000
7382     2000
4111     2000
2306     2000
5394     1000
5796     1000
6539     1000
10120    1000
7868     1000
7911     1000
8149     1000
8593     1000
4351     1000
8972     1000
9678     1000
9820     1000
9895     1000
12737    1000
12577    1000
10492    1000
12561    1000
12377    1000
12304    1000
12254    1000
12140    1000
12075    1000
12043    1000
12041    1000
11957    1000
11951    1000
11913    1000
11879    1000
11850    1000
11677    1000
11646    1000
11551    1000
11497    1000
11478    1000
11163    1000
10761    1000
10746    1000
10374    1000
Name: Wage, dtype: int64


У нас много одинаковых значений. Давайте посмотрим количество уникальных значений

## метод  value_counts()

In [47]:
nihonjin.value_counts()

1000    38
2000     7
3000     2
6000     1
4000     1
Name: Wage, dtype: int64

Преобразование в список методом $.tolist()$

In [49]:
nihonjin.value_counts().tolist()

[38, 7, 2, 1, 1]

In [58]:
football[(football.Age < football.Age.mean())&
        (football.Club == 'FC Barcelona')].Wage.mean()

77454.54545454546

# Задания

### 1. Какова средняя скорость (SprintSpeed) футболистов, зарплата (Wage) которых выше среднего? Ответ округлите до сотых. Ответ $67.56$
### 2. Какова средняя скорость (SprintSpeed) футболистов, зарплата (Wage) которых ниже среднего? Ответ округлите до сотых. Ответ $62.41$
### 3. Какую позицию (Position) занимает футболист с самой высокой зарплатой (Wage)?  Ответ $RF$
### 4. Сколько пенальти (Penalties) забили бразильские (Nationality, Brazil) футболисты за период, данные о котором представлены в датасете? Ответ $22789$
### 5. Укажите средний возраст (Age) игроков, у которых точность удара головой (HeadingAccuracy) > 50. Ответ округлите до сотых. Ответ $25.59$
### 6. Укажите возраст (Age) самого молодого игрока, у которого хладнокровие (Composure) и реакция (Reactions) превышают 90% от максимального значения, представленного в датасете. Ответ $24$
### 7. Определите, насколько средняя реакция (Reactions) самых взрослых игроков (т.е. игроков, чей возраст (Age) равен максимальному) больше средней реакции самых молодых игроков. Ответ округлите до сотых.  Ответ $22.64$
### 8. Из какой страны (Nationality) происходит больше всего игроков, чья стоимость (Value) превышает среднее значение?  Ответ $Spain$
### 9. Определите, во сколько раз средняя зарплата (Wage) голкипера (Position, GK) с максимальным значением показателя " Рефлексы" (GKReflexes) выше средней зарплаты голкипера с максимальным значением показателя "Владение мячом" (GKHandling). Ответ округлите до сотых. Ответ $2.77$
### 10. Определите, во сколько раз средняя сила удара (ShotPower) самых агрессивных игроков (игроков с максимальным значением показателя "Агрессивность" (Aggression)) выше средней силы удара игроков с минимальной агрессией. Ответ округлите до сотых. Ответ $2.08$

# Решение

### Задача 1

In [84]:
football[football.Wage>football.Wage.mean()].SprintSpeed.mean() # Задача 1

67.56638061585058

### Задача 2

In [83]:
football[football.Wage<football.Wage.mean()].SprintSpeed.mean() # Задача 2

62.41425430560645

### Задача 3

In [82]:
football[football.Wage == football.Wage.max()].Position         # Задача 3

0    RF
Name: Position, dtype: object

### Задача 4

In [81]:
football[football.Nationality == 'Brazil'].Penalties.sum()      # Задача 4

22789

### Задача 5

In [79]:
football[football.HeadingAccuracy > 50].Age.mean()              # Задача 5

25.58771113179944

### Задача 6

In [78]:
football[(football.Composure>0.9*football.Composure.max())&
         (football.Reactions>0.9*football.Reactions.max())].Age.min()  # Задача 6


24

### Задача 7

In [77]:
a = football[football.Age==football.Age.max()].Reactions            # Задача 7
b = football[football.Age==football.Age.min()].Reactions.mean()     # Задача 7
print(a-b)

1628    22.642857
Name: Reactions, dtype: float64


### Задача 8

In [62]:
# способ 1
series_national = football[football.Value > football.Value.mean()].Nationality
series_national

0                  Argentina
1                   Portugal
2                     Brazil
3                      Spain
4                    Belgium
                ...         
12736                England
12774    Republic of Ireland
12812                England
12857               China PR
12869                England
Name: Nationality, Length: 1747, dtype: object

In [63]:
dict_national = {}
for i in series_national:
    if i in dict_national:
        dict_national[i]+=1
    else:
        dict_national[i] = {}
        dict_national[i] = 1
max_nat=0
for i in dict_national:
    if dict_national[i] > max_nat:
        max_nat = dict_national[i]
        name = i
print(name)

Spain


In [74]:
# Способ 2
series_national.value_counts()#.loc['Spain']

211

### Задача 9

In [75]:
gk_1 = football[(football.Position == 'GK')&(football.GKReflexes == football.GKReflexes.max())].Wage
gk_2 = football[(football.Position == 'GK')&(football.GKHandling == football.GKHandling.max())].Wage.mean()
print(gk_1/gk_2)

3    2.765957
Name: Wage, dtype: float64


### Задача 10

In [76]:
ag_1 = football[football.Aggression==football.Aggression.max()].ShotPower.mean()
ag_2 = football[football.Aggression==football.Aggression.min()].ShotPower.mean()
print(ag_1/ag_2)

2.076923076923077


# Группировка данных. Метод .value_counts()

In [85]:
s = football['Nationality'].value_counts()
display(s)

England                 1368
Germany                  919
Spain                    671
France                   614
Argentina                574
                        ... 
Andorra                    1
Puerto Rico                1
Central African Rep.       1
Bermuda                    1
Guam                       1
Name: Nationality, Length: 156, dtype: int64

In [86]:
s.index

Index(['England', 'Germany', 'Spain', 'France', 'Argentina', 'Italy',
       'Colombia', 'Japan', 'Brazil', 'China PR',
       ...
       'Grenada', 'St Lucia', 'Lebanon', 'Nicaragua', 'Jordan', 'Andorra',
       'Puerto Rico', 'Central African Rep.', 'Bermuda', 'Guam'],
      dtype='object', length=156)

In [88]:
s.index[0]

'England'

In [89]:
s['England']

1368

In [93]:
s.loc[s > 300] # Страны в которых количество футболистов болшье 300 человек

England                1368
Germany                 919
Spain                   671
France                  614
Argentina               574
Italy                   469
Colombia                445
Japan                   425
Brazil                  423
China PR                349
Republic of Ireland     332
Saudi Arabia            324
Sweden                  319
Netherlands             309
Name: Nationality, dtype: int64

### Сколько футбольных клубов представлено в датасете?

In [94]:
s = football['Club'].value_counts()
display(s)

Shonan Bellmare      30
V-Varen Nagasaki     30
FC Emmen             29
Padova               29
De Graafschap        29
                     ..
Dijon FCO             8
Vitória Guimarães     7
Dinamo Zagreb         7
GD Chaves             7
Atlético Mineiro      6
Name: Club, Length: 650, dtype: int64

Всего разных клубов 650

### Нормализация данных относительно общего числа данных $12897$

In [96]:
s = football['Club'].value_counts(normalize=True)
display(s)

Shonan Bellmare      0.002326
V-Varen Nagasaki     0.002326
FC Emmen             0.002249
Padova               0.002249
De Graafschap        0.002249
                       ...   
Dijon FCO            0.000620
Vitória Guimarães    0.000543
Dinamo Zagreb        0.000543
GD Chaves            0.000543
Atlético Mineiro     0.000465
Name: Club, Length: 650, dtype: float64

### Группировка данных в диапазоне , методом .value_counts(bins=4), где 4 означает что данные разбиваем на 4 подгруппы то есть по 25%

In [99]:
s = football['Wage'].value_counts()
display(s)

1000      4601
2000      2652
3000      1596
4000       929
5000       566
          ... 
250000       1
185000       1
87000        1
300000       1
265000       1
Name: Wage, Length: 142, dtype: int64

In [101]:
s = football['Wage'].value_counts(bins=4)
display(s)

(435.999, 142000.0]     12818
(142000.0, 283000.0]       61
(283000.0, 424000.0]       16
(424000.0, 565000.0]        2
Name: Wage, dtype: int64

In [103]:
s.index[0]

Interval(435.999, 142000.0, closed='right')

In [104]:
s.index[0].right

142000.0

In [105]:
s.index[0].left

435.999

In [111]:
s.loc[s.index[0]]

12818

### Теперь видно, что распределение неравномерное и только 2 футболиста из 25 получают сверхвысокую зарплату.

In [114]:
football.loc[(football['Wage'] > s.index[3].left) & (football['Wage'] <= s.index[3].right)]

Unnamed: 0       Name  Age Nationality          Club      Value    Wage  \
0           0   L. Messi   31   Argentina  FC Barcelona  110500000  565000   
7           7  L. Suárez   31     Uruguay  FC Barcelona   80000000  455000   

  Position  Crossing  Finishing  ...  Penalties  Composure  Marking  \
0       RF        84         95  ...         75         96       33   
7       RS        77         93  ...         85         85       62   

   StandingTackle  SlidingTackle  GKDiving  GKHandling  GKKicking  \
0              28             26         6          11         15   
7              45             38        27          25         31   

   GKPositioning  GKReflexes  
0             14           8  
7             33          37  

[2 rows x 42 columns]

# Задание: В каких пределах находятся худшие 20% показателей точности ударов ногой (FKAccuracy)?

In [115]:
football['FKAccuracy'].value_counts(bins=5)

(21.2, 39.4]     4859
(39.4, 57.6]     3957
(57.6, 75.8]     2108
(2.908, 21.2]    1760
(75.8, 94.0]      213
Name: FKAccuracy, dtype: int64

# Преобразование в dataFrame( в таблицу). Метод .reset_index()

In [116]:
s = football['Nationality'].value_counts()
s_df = s.reset_index()
isplay(s_df)

index  Nationality
0                 England         1368
1                 Germany          919
2                   Spain          671
3                  France          614
4               Argentina          574
..                    ...          ...
151               Andorra            1
152           Puerto Rico            1
153  Central African Rep.            1
154               Bermuda            1
155                  Guam            1

[156 rows x 2 columns]

### Замена названия колонок

In [117]:
s_df.columns = ['Nationality','Players Count']
display(s_df)

Nationality  Players Count
0                 England           1368
1                 Germany            919
2                   Spain            671
3                  France            614
4               Argentina            574
..                    ...            ...
151               Andorra              1
152           Puerto Rico              1
153  Central African Rep.              1
154               Bermuda              1
155                  Guam              1

[156 rows x 2 columns]

# Решение задач:

1. У какого процента испанских футболистов (Nationality = 'Spain') зарплата (Wage) находится в пределах 25% минимума от наблюдаемого уровня зарплат среди испанских игроков? Ответ дайте в виде целого числа (округлите полученный результат) без знака %. Ответ $97$
2. Укажите количество уникальных сборных (Nationality), к которым относятся футболисты, выступающие за клуб (Club) "Manchester United". Ответ $13$
3. Oпределите двух футболистов из Бразилии (Nationality = 'Brazil'), выступающих за клуб (Club) 'Juventus'. Перечислите их имена (Name, как в датафрейме) через запятую в алфавитном порядке. Ответ $Alex Sandro,Douglas Costa$
4. Укажите, какой из клубов (Club) насчитывает большее количество футболистов возрастом (Age) старше 35 лет. Ответ $ Club Atlético Huracán, Hokkaido Consadole Sapporo $
5. С помощью функции value_counts с параметром bins разбейте всех футболистов сборной Аргентины (Nationality = 'Argentina') на 4 равные группы по возрасту. Сколько футболистов в возрасте от 34,75 до 41 года в сборной Аргентины? Ответ $49$
6. Сколько процентов футболистов из Испании (Nationality = 'Spain') имеют возраст (Age) 21 год? Введите с точностью до 2 знаков после запятой без указания знака % (например, 12.35). Ответ $11.77$


### Задача 1

In [120]:
s=football[football['Nationality']=='Spain'].Wage.value_counts(bins=4)
display(s)

(620.999, 95750.0]      651
(95750.0, 190500.0]      11
(190500.0, 285250.0]      6
(285250.0, 380000.0]      3
Name: Wage, dtype: int64

In [124]:
s[s.index[0]]/s.sum()

0.970193740685544

### Задача 2

In [118]:
football[football['Club']=='Manchester United'].Nationality.nunique()

13

### Задача 3

In [125]:
football[(football.Nationality == 'Brazil')&(football.Club=='Juventus')].Name

64      Alex Sandro
65    Douglas Costa
Name: Name, dtype: object

### Задача 4

In [126]:
football[football.Age>35].Club.value_counts()

Hokkaido Consadole Sapporo    4
Club Atlético Huracán         4
Club Atlético Tigre           3
San Martín de San Juan        3
Estudiantes de La Plata       3
                             ..
Deportes Iquique              1
Monarcas Morelia              1
Guadalajara                   1
Universidad de Chile          1
Randers FC                    1
Name: Club, Length: 212, dtype: int64

### Задача 5

In [127]:
football[football.Nationality=='Argentina'].Age.value_counts(bins=4)

(22.25, 28.5]      200
(15.974, 22.25]    197
(28.5, 34.75]      128
(34.75, 41.0]       49
Name: Age, dtype: int64

### Задача 6

In [129]:
football[(football.Nationality=='Spain')].Age.value_counts(normalize=True).head(5)

21    0.117735
22    0.095380
20    0.077496
23    0.071535
25    0.064083
Name: Age, dtype: float64

# Метод группировки данных $groupby$

### Посчитайте среднюю зарплату (Wage) и цену (Value) игроков разных позиций (Position). Представители какой позиции имеют самую высокую среднюю цену?

In [3]:
grouped_df = football.groupby(['Position'])[['Wage','Value']].mean().sort_values(['Value'],ascending=False)
display(grouped_df)

Wage         Value
Position                             
LF         75125.000000  2.736250e+07
RF        119000.000000  2.636583e+07
LAM        61000.000000  1.324583e+07
RAM        11800.000000  8.345000e+06
LS         15421.487603  5.848430e+06
LCM        13497.854077  4.318090e+06
RCM        13504.166667  4.217604e+06
RS         13482.456140  3.645833e+06
LW         12169.421488  3.603099e+06
RW         13822.784810  3.495738e+06
LDM        10243.243243  3.369459e+06
RCB        11270.334928  3.009055e+06
CAM         8688.783570  2.963199e+06
RDM        11282.608696  2.692717e+06
LCB         9753.056235  2.636516e+06
LM          7472.527473  2.466174e+06
ST          7852.513228  2.157252e+06
RM          6655.352480  2.138734e+06
LWB         5783.333333  1.931333e+06
CF          4862.745098  1.848627e+06
CDM         7121.562952  1.658900e+06
CM          5805.060918  1.617465e+06
RWB         6000.000000  1.555571e+06
GK          5352.833638  1.502502e+06
LB          6083.832335  1.429032e+06
CB          5328.581611  1.321251e+06
RB          5310.126582  1.248940e+06

In [5]:
grouped_df.loc['CB'] # Обращение к строки 

Wage     5.328582e+03
Value    1.321251e+06
Name: CB, dtype: float64

In [6]:
grouped_df.loc['CB']['Wage'] # Обращение к данной строке данного столбца

5328.581610833928

In [8]:
football.groupby(['Nationality'])[['Wage','Age','ShotPower']].mean().sort_values(['Wage'],ascending=False).head(10)

Wage        Age  ShotPower
Nationality                                           
Dominican Republic  71000.000000  23.000000  75.500000
Egypt               35545.454545  25.818182  59.363636
Gabon               28900.000000  26.400000  56.900000
Croatia             26722.222222  24.819444  54.305556
Equatorial Guinea   25666.666667  28.000000  55.333333
Belgium             20024.390244  24.030488  56.390244
Ecuador             18333.333333  24.619048  60.666667
Uruguay             17590.361446  26.771084  56.192771
Brazil              17371.158392  27.898345  58.203310
Algeria             15810.810811  27.027027  56.567568

## Доминиканская республика имеет самую высoкую зарплату у футболистов ?

In [10]:
football.loc[football['Nationality'] == 'Dominican Republic'][['Name','Club','Wage','Age','ShotPower']]

Name         Club    Wage  Age  ShotPower
291      Mariano  Real Madrid  140000   24         85
6279  L. Quezada   Córdoba CF    2000   22         66

## параметр - подсчет уникальных значений .nunique()

In [12]:
football.groupby(['Nationality'])[['Club','Name']].nunique().sort_values('Club',ascending = False)

Club  Name
Nationality                     
Brazil                 173   408
France                 146   608
England                138  1294
Argentina              133   554
Spain                  117   650
...                    ...   ...
Central African Rep.     1     1
Jordan                   1     1
Rwanda                   1     1
Puerto Rico              1     1
Nicaragua                1     1

[156 rows x 2 columns]

# Задания
## 1. С помощью функции groupby посчитайте сумму зарплат (Wage) футболистов клуба (Club) "Chelsea". Ответ $2035000$
## 2. Определите максимальную зарплату футболиста национальности (Nationality) Аргентина ("Argentina") в возрасте 30 лет.  Ответ $300000$
## 3. Определите минимальную зарплату самого младшего футболиста национальности (Nationality) Аргентина ("Argentina").  Ответ $1000$
## 4. Определите максимальную силу (Strength) и баланс (Balance) среди игроков клуба (Club) "FC Barcelona" из Аргентины ("Argentina").   Ответ $59;95$

### Задача 1

In [16]:
grouped_df = football.groupby(['Club'])['Wage'].sum()
display(grouped_df['Chelsea'])

2035000

### Задача 2

In [18]:
grouped_df = football[football.Age==30].groupby(['Nationality']).Wage.max()
display(grouped_df['Argentina'])

300000

### Задача 3

In [22]:
grouped_df = football[football.Age==football.Age.min()].groupby(['Nationality']).Wage.min()
display(grouped_df['Argentina'])

1000

### Задача 4

In [26]:
grouped_df = football[football.Nationality=='Argentina'].groupby(['Club'])[['Strength','Balance']].max()
print(grouped_df)
print('------')
grouped_df.loc['FC Barcelona']['Strength']

                   Strength  Balance
Club                                
AD Alcorcón              85       38
Ajax                     63       79
Al Fayha                 61       79
Alanyaspor               43       84
Alianza Petrolera        72       54
...                     ...      ...
Unión de Santa Fe        78       81
Valencia CF              88       34
VfB Stuttgart            77       87
Vélez Sarsfield          82       76
West Ham United          73       89

[133 rows x 2 columns]
------


59

# Метод $.pivot\_table()$ для группировки, аналог $.groupby()$

In [29]:
pivot = football.loc[football['Club'].isin(['FC Barcelona','Real Madrid','Juventus',
                                            'Manchester United'])].pivot_table(values=['Wage'],
index=['Nationality'],
columns=['Club'],
aggfunc='sum',
margins=True)
display(pivot)

Wage                                           \
Club               FC Barcelona   Juventus Manchester United Real Madrid   
Nationality                                                                
Argentina              565000.0   205000.0          206000.0         NaN   
Belgium                110000.0        NaN          230000.0    240000.0   
Bosnia Herzegovina          NaN   180000.0               NaN         NaN   
Brazil                 465000.0   335000.0           91000.0    570000.0   
Chile                  205000.0        NaN          215000.0         NaN   
Costa Rica                  NaN        NaN               NaN    195000.0   
Croatia                260000.0   160000.0               NaN    420000.0   
Dominican Republic          NaN        NaN               NaN    140000.0   
Ecuador                     NaN        NaN          120000.0         NaN   
England                     NaN        NaN          478000.0         NaN   
France                 495000.0   145000.0          375000.0    459000.0   
Germany                240000.0   160000.0               NaN    355000.0   
Italy                       NaN   460000.0           88000.0         NaN   
Ivory Coast                 NaN        NaN          105000.0         NaN   
Morocco                     NaN   160000.0               NaN         NaN   
Netherlands            135000.0        NaN               NaN         NaN   
Poland                      NaN   125000.0               NaN         NaN   
Portugal               125000.0   405000.0               NaN         NaN   
Scotland                    NaN        NaN           11000.0         NaN   
Serbia                      NaN        NaN          165000.0         NaN   
Spain                  912000.0        NaN          260000.0   1404000.0   
Uruguay                455000.0        NaN               NaN         NaN   
Wales                       NaN        NaN           13000.0    355000.0   
All                   3967000.0  2335000.0         2357000.0   4138000.0   

                              
Club                     All  
Nationality                   
Argentina             976000  
Belgium               580000  
Bosnia Herzegovina    180000  
Brazil               1461000  
Chile                 420000  
Costa Rica            195000  
Croatia               840000  
Dominican Republic    140000  
Ecuador               120000  
England               478000  
France               1474000  
Germany               755000  
Italy                 548000  
Ivory Coast           105000  
Morocco               160000  
Netherlands           135000  
Poland                125000  
Portugal              530000  
Scotland               11000  
Serbia                165000  
Spain                2576000  
Uruguay               455000  
Wales                 368000  
All                 12797000

In [31]:
pivot = football.loc[football['Club'].isin(['FC Barcelona','Real Madrid','Juventus',
                                'Manchester United'])].pivot_table(values=['Wage'],
index=['Nationality'],
columns=['Club'],
aggfunc='sum',
margins=True,
fill_value=0)
display(pivot)

Wage                                         \
Club               FC Barcelona Juventus Manchester United Real Madrid   
Nationality                                                              
Argentina                565000   205000            206000           0   
Belgium                  110000        0            230000      240000   
Bosnia Herzegovina            0   180000                 0           0   
Brazil                   465000   335000             91000      570000   
Chile                    205000        0            215000           0   
Costa Rica                    0        0                 0      195000   
Croatia                  260000   160000                 0      420000   
Dominican Republic            0        0                 0      140000   
Ecuador                       0        0            120000           0   
England                       0        0            478000           0   
France                   495000   145000            375000      459000   
Germany                  240000   160000                 0      355000   
Italy                         0   460000             88000           0   
Ivory Coast                   0        0            105000           0   
Morocco                       0   160000                 0           0   
Netherlands              135000        0                 0           0   
Poland                        0   125000                 0           0   
Portugal                 125000   405000                 0           0   
Scotland                      0        0             11000           0   
Serbia                        0        0            165000           0   
Spain                    912000        0            260000     1404000   
Uruguay                  455000        0                 0           0   
Wales                         0        0             13000      355000   
All                     3967000  2335000           2357000     4138000   

                              
Club                     All  
Nationality                   
Argentina             976000  
Belgium               580000  
Bosnia Herzegovina    180000  
Brazil               1461000  
Chile                 420000  
Costa Rica            195000  
Croatia               840000  
Dominican Republic    140000  
Ecuador               120000  
England               478000  
France               1474000  
Germany               755000  
Italy                 548000  
Ivory Coast           105000  
Morocco               160000  
Netherlands           135000  
Poland                125000  
Portugal              530000  
Scotland               11000  
Serbia                165000  
Spain                2576000  
Uruguay               455000  
Wales                 368000  
All                 12797000

# Если в данных есть пропуски, то метод $.groupby()$ не сработет без предварительной чистки

In [32]:
grouped_df = football.groupby(['Nationality'])[['FC Barcelona','Real Madrid','Juventus',
                                'Manchester United']].Wage.sum()
print(grouped_df)

KeyError: "Columns not found: 'FC Barcelona', 'Manchester United', 'Real Madrid', 'Juventus'"

In [34]:
pivot.loc['Argentina']['Wage']['Manchester United']

206000

In [35]:
pivot.loc['Argentina']

      Club             
Wage  FC Barcelona         565000
      Juventus             205000
      Manchester United    206000
      Real Madrid               0
      All                  976000
Name: Argentina, dtype: int64

In [36]:
pivot.loc['Argentina']['Wage']['All']

976000

In [37]:
pivot.loc['All']['Wage']['Juventus']

2335000

In [40]:
pivot.index

Index(['Argentina', 'Belgium', 'Bosnia Herzegovina', 'Brazil', 'Chile',
       'Costa Rica', 'Croatia', 'Dominican Republic', 'Ecuador', 'England',
       'France', 'Germany', 'Italy', 'Ivory Coast', 'Morocco', 'Netherlands',
       'Poland', 'Portugal', 'Scotland', 'Serbia', 'Spain', 'Uruguay', 'Wales',
       'All'],
      dtype='object', name='Nationality')

In [41]:
pivot.index[0]

'Argentina'

In [42]:
pivot.loc[pivot.index[0]]['Wage']['All']

976000

# Пример: Создайте сводную таблицу, содержащую сведения о количестве игроков, занимающих разные позиции в каждом клубе. Отсутствующие значения замените нулями. Каково среднее количество вратарей (GK) в клубе? Ответ округлите до трёх цифр после запятой.

In [44]:

pivot = football.pivot_table(values=['Name'],
index=['Position'],
columns=['Club'],
aggfunc='count',
margins=True,
fill_value=0)
display(pivot.head())

Name                                             \
Club      SSV Jahn Regensburg 1. FC Heidenheim 1846 1. FC Kaiserslautern   
Position                                                                   
CAM                         0                     1                    0   
CB                          5                     3                    3   
CDM                         1                     0                    3   
CF                          0                     0                    0   
CM                          2                     1                    0   

                                                                       \
Club     1. FC Köln 1. FC Magdeburg 1. FC Nürnberg 1. FC Union Berlin   
Position                                                                
CAM               0               0              1                  1   
CB                1               4              1                  2   
CDM               1               0              2                  2   
CF                0               1              0                  0   
CM                2               0              1                  1   

                                                ...                   \
Club     1. FSV Mainz 05 AC Ajaccio AC Horsens  ... Yeni Malatyaspor   
Position                                        ...                    
CAM                    0          3          0  ...                2   
CB                     2          1          3  ...                2   
CDM                    0          0          1  ...                3   
CF                     0          0          0  ...                0   
CM                     1          1          2  ...                2   

                                                                            \
Club     Yeovil Town Yokohama F. Marinos Zagłębie Lubin Zagłębie Sosnowiec   
Position                                                                     
CAM                0                   3              0                  3   
CB                 3                   4              0                  2   
CDM                0                   2              2                  2   
CF                 0                   0              0                  0   
CM                 1                   1              1                  0   

                                                                      
Club     Çaykur Rizespor Örebro SK Östersunds FK Śląsk Wrocław   All  
Position                                                              
CAM                    1         0             1             0   633  
CB                     2         4             1             1  1403  
CDM                    5         0             0             3   691  
CF                     0         0             1             0    51  
CM                     0         1             3             1  1067  

[5 rows x 651 columns]

In [50]:
pivot.loc['GK']['Name'].mean()#/2#/len(name_list)

5.04147465437788

In [51]:
pivot.loc['GK']['Name'].sum()/len(pivot.loc['GK']['Name'])

5.04147465437788

## Используя таблицу, созданную на предыдущем шаге, определите, сколько клубов не содержат данных о центральных полузащитниках. (CM)

In [54]:
pivot.loc['CM']['Name'].value_counts()

1       211
2       149
0       136
3        89
4        39
5        21
6         5
1067      1
Name: CM, dtype: int64

In [55]:
pivot.loc['CM']['Name'].value_counts()[0] # указывая в скобках имя строки , в данном случае 0

136

# Задачи
## 1. С помощью сводной таблицы и функции loc посчитайте, сколько получают ("Wage") русские футболисты ("Russia"), играющие за ФК "AS Monaco". Ответ $61000$
## 2. Создайте сводную таблицу, содержащую сведения о средней скорости футболистов, занимающих разные позиции (Position) в разных футбольных клубах (Club). Выпишите первые три позиции игроки котороые обладют наибольше скоростью. Ответ $RS-RM-LM$
## 3. Используя таблицу, созданную на предыдущем шаге, отметьте названия трёх клубов, в которых центральные форварды (ST) обладают наибольшей средней скоростью. Ответ $Vitória-Sydney FC-CD Lugo$
## 4. Составьте сводную таблицу по трем клубам $'FC Barcelona','Real Madrid','Manchester United'$ количества игроков, возраст которых больше среднего возраста 

### Задача 1

In [56]:
pivot = football.pivot_table(values=['Wage'],
index=['Nationality'],
columns=['Club'],
aggfunc='mean',
margins=True,
fill_value=0)
display(pivot)

Wage                                             \
Club         SSV Jahn Regensburg 1. FC Heidenheim 1846 1. FC Kaiserslautern   
Nationality                                                                   
Afghanistan                  0.0                     0             0.000000   
Albania                      0.0                     0             0.000000   
Algeria                      0.0                     0             0.000000   
Andorra                      0.0                     0             0.000000   
Angola                       0.0                     0             0.000000   
...                          ...                   ...                  ...   
Venezuela                    0.0                     0             0.000000   
Wales                        0.0                     0             0.000000   
Zambia                       0.0                     0             0.000000   
Zimbabwe                     0.0                     0             0.000000   
All                       3750.0                  3800          1434.782609   

                                                                          \
Club        1. FC Köln 1. FC Magdeburg 1. FC Nürnberg 1. FC Union Berlin   
Nationality                                                                
Afghanistan        0.0             0.0       0.000000           0.000000   
Albania            0.0             0.0       0.000000           0.000000   
Algeria            0.0             0.0       0.000000           0.000000   
Andorra            0.0             0.0       0.000000           0.000000   
Angola             0.0             0.0       0.000000           0.000000   
...                ...             ...            ...                ...   
Venezuela          0.0             0.0       0.000000           0.000000   
Wales              0.0             0.0       0.000000           0.000000   
Zambia             0.0             0.0       0.000000           0.000000   
Zimbabwe           0.0             0.0       0.000000           0.000000   
All             9200.0          4000.0    7153.846154        8545.454545   

                                                     ...                   \
Club        1. FSV Mainz 05   AC Ajaccio AC Horsens  ... Yeni Malatyaspor   
Nationality                                          ...                    
Afghanistan        0.000000     0.000000          0  ...         0.000000   
Albania            0.000000  2000.000000          0  ...         0.000000   
Algeria            0.000000     0.000000          0  ...         0.000000   
Andorra            0.000000     0.000000          0  ...         0.000000   
Angola             0.000000     0.000000          0  ...         0.000000   
...                     ...          ...        ...  ...              ...   
Venezuela          0.000000     0.000000          0  ...         0.000000   
Wales              0.000000     0.000000          0  ...         0.000000   
Zambia             0.000000     0.000000          0  ...         0.000000   
Zimbabwe           0.000000     0.000000          0  ...         0.000000   
All            11363.636364  2210.526316       2200  ...      4318.181818   

                                                             \
Club         Yeovil Town Yokohama F. Marinos Zagłębie Lubin   
Nationality                                                   
Afghanistan     0.000000                   0              0   
Albania         0.000000                   0              0   
Algeria      1000.000000                   0              0   
Andorra         0.000000                   0              0   
Angola          0.000000                   0              0   
...                  ...                 ...            ...   
Venezuela       0.000000                   0              0   
Wales        2000.000000                   0              0   
Zambia          0.000000                   0              0   
Zimbabwe        0.000

In [57]:
pivot.loc['Russia']['Wage']['AS Monaco']

61000.0

### Задача 2

In [58]:
df2 = football.pivot_table(columns = 'Position', index = 'Club',
                           values = 'SprintSpeed', aggfunc = 'mean').max().sort_values(ascending=False)
display(df2)

Position
RS     94.0
RM     94.0
LM     94.0
LS     93.0
RB     93.0
LB     93.0
CF     91.0
LW     91.0
ST     91.0
RDM    91.0
RW     91.0
RCM    90.0
CAM    89.0
LF     88.0
RCB    88.0
LCB    88.0
RWB    86.0
RF     86.0
CM     86.0
LWB    86.0
LAM    85.0
LCM    84.0
CDM    83.0
LDM    79.0
RAM    78.0
CB     78.0
GK     62.0
dtype: float64

### Задача 3

In [59]:
df2 = football.pivot_table(columns = 'Position', index = 'Club', values = 'SprintSpeed', aggfunc = 'mean')
display(df2['ST'].sort_values(ascending=False))

Club
Vitória            91.000000
Sydney FC          88.000000
CD Lugo            87.000000
Brøndby IF         84.666667
Leicester City     84.500000
                     ...    
Valencia CF              NaN
VfL Bochum 1848          NaN
VfL Wolfsburg            NaN
Waterford FC             NaN
Wigan Athletic           NaN
Name: ST, Length: 650, dtype: float64

In [68]:
football[football.Age>41]

Unnamed: 0           Name  Age        Nationality                Club  \
1628         1695       O. Pérez   45             Mexico             Pachuca   
5394         5556    S. Narazaki   42              Japan      Nagoya Grampus   
12422       12651      T. Warner   44  Trinidad & Tobago  Accrington Stanley   
12873       13102  K. Pilkington   44            England    Cambridge United   

         Value  Wage Position  Crossing  Finishing  ...  Penalties  Composure  \
1628   1300000  8000       GK        19         14  ...         23         62   
5394     40000  1000       GK        23         11  ...         11         45   
12422        0  1000       GK        19         13  ...         17         46   
12873        0  1000       GK        11         12  ...         22         56   

       Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  \
1628        23              12             11        70          64   
5394        27              24             22        64          63   
12422       19              15             14        48          56   
12873       15              15             13        45          48   

       GKKicking  GKPositioning  GKReflexes  
1628          65             73          74  
5394          62             67          63  
12422         56             60          44  
12873         44             49          46  

[4 rows x 42 columns]

### Задача 4

In [71]:
pivot = football[football.Age>football.Age.mean()].loc[football['Club'].isin(['FC Barcelona','Real Madrid',
                                'Manchester United'])].pivot_table(values=['Name'],
index=['Nationality'],
columns=['Club'],
aggfunc='count',
margins=True)
display(pivot)

Name                                  
Club        FC Barcelona Manchester United Real Madrid All
Nationality                                               
Argentina            1.0               2.0         NaN   3
Belgium              1.0               1.0         1.0   3
Brazil               1.0               NaN         2.0   3
Chile                1.0               1.0         NaN   2
Costa Rica           NaN               NaN         1.0   1
Croatia              1.0               NaN         1.0   2
Ecuador              NaN               1.0         NaN   1
England              NaN               3.0         NaN   3
France               NaN               1.0         2.0   3
Germany              1.0               NaN         1.0   2
Italy                NaN               1.0         NaN   1
Netherlands          1.0               NaN         NaN   1
Serbia               NaN               1.0         NaN   1
Spain                3.0               1.0         3.0   7
Uruguay              1.0               NaN         NaN   1
Wales                NaN               NaN         1.0   1
All                 11.0              12.0        12.0  35

# Предобработка данных

«Сырые» данные, с которыми работает аналитик, всегда нужно предобработать. Данные в исходном виде практически всегда содержат:

        ошибки ввода,
        технические ошибки,
        пропущенные значения,
        дубликаты,
        информационный шум и ненужную информацию.

Все эти причины не позволяют извлечь полезную информацию из данных, поэтому предобработка — важная часть работы аналитика и data scientist'а, занимающая порой больше времени, чем сам анализ. Предоработку данных можно сравнить с подготовкой к длинному путешествию: покупка билетов и бронь отеля, маршруты, сбор вещей — всё это скучная, но важная часть путешествия. Если её сделать неправильно и не вовремя, то само путешествие может провалиться

In [1]:
import pandas as pd # импорт библиотеки Пандас

Параметры pd.read_csv

        header = None — загрузить без строки с заголовком,
        skiprows=n — пропустить n строк (часто у документов бывает техническая «шапка»),
        encoding — загрузить в конкретной кодировке,
        na_values — список значений, который нужно заменить на NaN (специальный объект, обозначающий
        пропущенное  значение).

In [3]:
users = pd.read_csv('users.csv',sep='\t',encoding='KOI8-R')
users.columns =['user_id','email','geo']
users

user_id                     email          geo
0   User_943       Accumanst@gmail.com       Ижевск
1   User_908         Advismowr@mail.ru       Ижевск
2   User_962           Anachso@ukr.net    Краснодар
3   User_973          Antecia@inbox.ru        Пермь
4   User_902        Balliaryva@ukr.net          NaN
..       ...                       ...          ...
95  User_959  UpdatesCurious@yahoo.com       Тюмень
96  User_901        V2artierso@mail.ru  Арзангелтск
97  User_970          Vashoterlo@bk.ru      Воронеж
98  User_965       Visuareda@yahoo.com      Воронеж
99  User_921              Aavast@ya.ru       Ижевск

[100 rows x 3 columns]

In [9]:
log = pd.read_csv("log.csv",header=None)
log.columns = ['user_id','time', 'bet','win']
log

user_id                  time      bet      win
0    Запись пользователя № - user_919  [2019-01-01 14:06:51      NaN      NaN
1    Запись пользователя № - user_973  [2019-01-01 14:51:16      NaN      NaN
2    Запись пользователя № - user_903  [2019-01-01 16:31:16      NaN      NaN
3    Запись пользователя № - user_954  [2019-01-01 17:17:51      NaN      NaN
4    Запись пользователя № - user_954  [2019-01-01 21:31:18      NaN      NaN
..                                ...                   ...      ...      ...
995  Запись пользователя № - user_984   [2019-04-20 9:59:58   9754.0      NaN
996                            #error                   NaN  10054.0  29265.0
997                            #error                   NaN  10454.0      NaN
998                            #error                   NaN   1000.0      NaN
999                            #error                   NaN  10754.0      NaN

[1000 rows x 4 columns]

# Функция $query$

### $Pandas$ помогает фильтровать значения без создания циклов, главное — сформулировать, что именно ты ищешь. Наиболее мощный и удобный способ проводить фильтрацию — использовать функцию $query$. Она позволяет писать сложные запросы к датафрейму.

In [4]:
sample = pd.read_csv('sample.csv')
sample.query('Age>20')

Name         City  Age Profession
0         Иванов       Москва   25     Клинер
1            NaN    Волгоград   31   Менеджер
2         Иванов       Москва   25     Клинер
3        sidorov  Владивосток   43   Менеджер
4        _______        Курск   50   Водитель
6          Сажин       Сургут   29    Рабочий
7   Котлеревский   Йошкар-Ола   32    Рабочий
8     Завалишина    Чебоксары   36   Менеджер
9         Левина          NaN   55    Рабочий
10          Фикс         Рига   42    Рабочий

In [5]:
sample.Age.min()

19

In [6]:
sample.query('City in ["Рига","Сочи"]')

Name  City  Age Profession
5   Кузнецов  Сочи   19    Рабочий
10      Фикс  Рига   42    Рабочий

In [7]:
sample.query('City in ["Рига", "Сочи","Чебоксары", "Сургут"] & 21<Age<50 & Profession!="Менеджер"')

Name    City  Age Profession
6   Сажин  Сургут   29    Рабочий
10   Фикс    Рига   42    Рабочий

In [10]:
log2=log.query('bet<2000 & win>0')
log2

user_id                  time     bet     win
151  Запись пользователя № - user_982  [2019-01-16 21:54:22   100.0  4749.0
189  Запись пользователя № - user_964  [2019-01-21 18:34:44   200.0  4667.0
205  Запись пользователя № - user_931   [2019-01-22 5:26:59   300.0  4319.0
232  Запись пользователя № - user_998   [2019-01-25 8:57:20   500.0  5069.0
265  Запись пользователя № - user_998  [2019-01-29 10:37:55   500.0  6294.0
..                                ...                   ...     ...     ...
962  Запись пользователя № - user_976  [2019-04-19 14:18:45   400.0  5997.0
965  Запись пользователя № - user_900  [2019-04-19 19:31:48   900.0  6767.0
967  Запись пользователя № - user_975  [2019-04-19 22:25:15  1000.0  6108.0
981                            #error                   NaN   800.0  7035.0
991  Запись пользователя № - user_965  [2019-04-20 12:55:41   800.0  6927.0

[113 rows x 4 columns]

# Функции $str.match$ и $str.contains$

Часто приходится фильтровать значения в столбце по соответствию строки или её части. Для этого есть две полезные функции:

    str.match ("abc") — ищет строки, которые начинаются c abc,
    str.contains ("abc") — ищет строки, в которых есть abc.
   
Обе функции применимы только к объектам Series и не могут работать с NaN, необходимо использовать параметр na = False

В результате применения функции мы получаем своеобразную маску, которую мы можем наложить на датафрейм для фильтрации.

In [11]:
sample

Name         City  Age Profession
0         Иванов       Москва   25     Клинер
1            NaN    Волгоград   31   Менеджер
2         Иванов       Москва   25     Клинер
3        sidorov  Владивосток   43   Менеджер
4        _______        Курск   50   Водитель
5       Кузнецов         Сочи   19    Рабочий
6          Сажин       Сургут   29    Рабочий
7   Котлеревский   Йошкар-Ола   32    Рабочий
8     Завалишина    Чебоксары   36   Менеджер
9         Левина          NaN   55    Рабочий
10          Фикс         Рига   42    Рабочий

In [12]:
sample.Name.str.match("К", na=False)

0     False
1     False
2     False
3     False
4     False
5      True
6     False
7      True
8     False
9     False
10    False
Name: Name, dtype: bool

In [13]:
sample[sample.Name.str.match("К", na=False)]

Name        City  Age Profession
5      Кузнецов        Сочи   19    Рабочий
7  Котлеревский  Йошкар-Ола   32    Рабочий

### Если хотим получить значения которые не удолетворяют условю:

In [14]:
sample[~sample.Name.str.match("К", na=False)]

Name         City  Age Profession
0       Иванов       Москва   25     Клинер
1          NaN    Волгоград   31   Менеджер
2       Иванов       Москва   25     Клинер
3      sidorov  Владивосток   43   Менеджер
4      _______        Курск   50   Водитель
6        Сажин       Сургут   29    Рабочий
8   Завалишина    Чебоксары   36   Менеджер
9       Левина          NaN   55    Рабочий
10        Фикс         Рига   42    Рабочий

# Простые задачи на закрепление

1. Найдите записи в sample, где в городах есть буква «о».
2. Сохраните в переменную new_log датафрейм, из которого удалены записи с ошибкой в поле user_id
3. Оставьте в new_log только те данные, в которых есть значения bet и win, используя функцию query

### Задача 1

In [16]:
sample[sample.City.str.contains("о",na=False)]

Name         City  Age Profession
0        Иванов       Москва   25     Клинер
1           NaN    Волгоград   31   Менеджер
2        Иванов       Москва   25     Клинер
3       sidorov  Владивосток   43   Менеджер
5      Кузнецов         Сочи   19    Рабочий
7  Котлеревский   Йошкар-Ола   32    Рабочий
8    Завалишина    Чебоксары   36   Менеджер

### Задача 2

In [ ]:
new_log=log[~log.user_id.str.match("#error",na=False)]
new_log

### Задача 3

In [20]:
new_log.query('win>0 & bet>0')

user_id                  time        bet  \
14   Запись пользователя № - user_917   [2019-01-02 8:57:36   145732.0   
29   Запись пользователя № - user_942  [2019-01-04 13:59:42  1678321.0   
151  Запись пользователя № - user_982  [2019-01-16 21:54:22      100.0   
189  Запись пользователя № - user_964  [2019-01-21 18:34:44      200.0   
205  Запись пользователя № - user_931   [2019-01-22 5:26:59      300.0   
..                                ...                   ...        ...   
967  Запись пользователя № - user_975  [2019-04-19 22:25:15     1000.0   
971  Запись пользователя № - user_912  [2019-04-20 10:35:49    10554.0   
972  Запись пользователя № - user_926  [2019-04-20 10:35:50    10354.0   
976  Запись пользователя № - user_970  [2019-04-20 10:35:54    10354.0   
991  Запись пользователя № - user_965  [2019-04-20 12:55:41      800.0   

           win  
14   1987653.0  
29   9876543.0  
151     4749.0  
189     4667.0  
205     4319.0  
..         ...  
967     6108.0  
971    31799.0  
972    30244.0  
976    30691.0  
991     6927.0  

[133 rows x 4 columns]

# Преобразование данных

Часто данные оказываются плохо приспособленными к анализу, поэтому их следует преобразовать. Для решения этой задачи используется функция apply.  

df.column_name.apply(func)

Apply вызывает свою функцию для каждого элемента колонки. Функции, с которыми работает apply, бывают двух видов:

    лямбда-функции (можно задать внутри apply),
    обычные именные функции (определяются через def).


In [21]:
sample.Age.apply(lambda x:x**2)

0      625
1      961
2      625
3     1849
4     2500
5      361
6      841
7     1024
8     1296
9     3025
10    1764
Name: Age, dtype: int64

In [22]:
# Обнулим значения возраста для всех возрастов больше 19
def func(x):
    if x<20:
        return x
    else:
        return 0
sample.Age.apply(func)
# Передаем только имя функции

0      0
1      0
2      0
3      0
4      0
5     19
6      0
7      0
8      0
9      0
10     0
Name: Age, dtype: int64

# Задания для закрепления

1. 

In [25]:
sample

Name         City  Age Profession
0         Иванов       Москва   25     Клинер
1            NaN    Волгоград   31   Менеджер
2         Иванов       Москва   25     Клинер
3        sidorov  Владивосток   43   Менеджер
4        _______        Курск   50   Водитель
5       Кузнецов         Сочи   19    Рабочий
6          Сажин       Сургут   29    Рабочий
7   Котлеревский   Йошкар-Ола   32    Рабочий
8     Завалишина    Чебоксары   36   Менеджер
9         Левина          NaN   55    Рабочий
10          Фикс         Рига   42    Рабочий

In [30]:
sample2=sample.Age.apply(lambda x: x+1)
sample[sample.Age=sample2]

SyntaxError: invalid syntax (<ipython-input-30-b79b3b79a6ee>, line 2)